### PR TITLE
Back to vertx

### PR DESCRIPTION
**Pull request by Sammers21**
This pull request removes RIO and adds Vert.x as the implementation for `FileSystem`

This changes will improve `FileSystem` performance:

Benchmarks:

**AWS t3a.2xlarge 8xCPU 32GB RAM**

```bash
$ uname -a
Linux ip-172-31-12-154 5.3.0-1034-aws #36-Ubuntu SMP Tue Aug 18 08:58:43 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
```

RIO:
```bash
$ git checkout rio_repeat_100
$ mvn clean install -Dtest=FileStorageTest#saveAndLoadHugeFiles
Run #0. Time: 1s160ms
Run #1. Time: 8s82ms
Run #2. Time: 8s31ms
Run #3. Time: 8s41ms
Run #4. Time: 8s41ms
Run #5. Time: 7s972ms
Run #6. Time: 8s57ms
Run #7. Time: 8s43ms
Run #8. Time: 8s28ms
Run #9. Time: 8s43ms
Run #10. Time: 8s29ms
Run #11. Time: 8s69ms
Run #12. Time: 8s31ms
Run #13. Time: 8s59ms
Run #14. Time: 8s41ms
Run #15. Time: 8s48ms
Run #16. Time: 8s28ms
Run #17. Time: 8s31ms
Run #18. Time: 8s30ms
Run #19. Time: 8s32ms
```

Vert.x:

```bash
git checkout back_to_vertx_repeat_100
mvn clean install -Dtest=FileStorageTest#saveAndLoadHugeFiles
Run #0. Time: 1s993ms
Run #1. Time: 1s335ms
Run #2. Time: 1s187ms
Run #3. Time: 1s294ms
Run #4. Time: 1s455ms
Run #5. Time: 1s330ms
Run #6. Time: 1s379ms
Run #7. Time: 1s146ms
Run #8. Time: 1s412ms
Run #9. Time: 1s226ms
Run #10. Time: 1s180ms
Run #11. Time: 1s382ms
Run #12. Time: 1s352ms
Run #13. Time: 1s670ms
Run #14. Time: 1s431ms
Run #15. Time: 1s131ms
Run #16. Time: 1s397ms
Run #17. Time: 1s115ms
Run #18. Time: 1s199ms
Run #19. Time: 1s342ms
```

On a storage-optimised AWS instance **i3.large** running **Ubuntu 18.04**  the benchmarks results are the following:

**RIO**:
```
[INFO] Running com.artipie.asto.FileStorageTest
Run #0. Time: 1s605ms
Run #1. Time: 20s617ms
Run #2. Time: 20s515ms
Run #3. Time: 20s602ms
Run #4. Time: 20s428ms
Run #5. Time: 20s599ms
Run #6. Time: 20s376ms
Run #7. Time: 20s445ms
Run #8. Time: 20s332ms
Run #9. Time: 20s487ms
Run #10. Time: 20s441ms
Run #11. Time: 20s454ms
Run #12. Time: 20s335ms
Run #13. Time: 20s463ms
Run #14. Time: 20s444ms
Run #15. Time: 20s443ms
Run #16. Time: 20s435ms
Run #17. Time: 20s439ms
Run #18. Time: 20s305ms
Run #19. Time: 20s494ms
```

**Vert.x**:
```
Run #0. Time: 4s95ms
Run #1. Time: 2s228ms
Run #2. Time: 2s142ms
Run #3. Time: 2s119ms
Run #4. Time: 2s77ms
Run #5. Time: 2s157ms
Run #6. Time: 2s115ms
Run #7. Time: 2s124ms
Run #8. Time: 2s32ms
Run #9. Time: 2s40ms
Run #10. Time: 2s193ms
Run #11. Time: 2s101ms
Run #12. Time: 2s104ms
Run #13. Time: 2s122ms
Run #14. Time: 2s77ms
Run #15. Time: 2s158ms
Run #16. Time: 2s285ms
Run #17. Time: 2s210ms
Run #18. Time: 2s440ms
Run #19. Time: 2s524ms
```

### CODE REVIEW

**Comment by olegmoz** [comment]
@Sammers21 there is no proper description and no related issue for this PR. The PR contains a lot of changes unrelated to each other

Please create proper PR if you want changes to be reviewed and accepted

**Comment by 0crat** [comment]
Job `gh:artipie/asto#279` is not assigned, can't get performer

**Comment by 0crat** [comment]
Pull request [#279](https://github.com/artipie/asto/issues/279) was not merged, no payment for `ARC`, see [§28](http://www.zerocracy.com/policy.html#28)

**Comment by olegmoz** [comment]
@Sammers21 please request review only when everything is ready: only required changes are in PR and build is successful

File: src/main/java/com/artipie/asto/fs/FileStorage.java [hunk 1]
```diff
@@ -68,54 +64,34 @@
...
+    public FileStorage(final Path path, final FileSystem fls) {
         this.dir = path;
-        this.exec = exec;
+        this.fls = fls;
     }
 
     @Override
     public CompletableFuture<Boolean> exists(final Key key) {
-        return CompletableFuture.supplyAsync(
+        return Single.fromCallable(
```
Comment (olegmoz) [inline src/main/java/com/artipie/asto/fs/FileStorage.java:84]: @Sammers21 what is the reason for this change?

File: src/main/java/com/artipie/asto/fs/FileStorage.java [hunk 2]
```diff
@@ -68,54 +64,34 @@
...
-            this.exec
-        );
+            }
+        ).to(SingleInterop.get()).toCompletableFuture();
     }
 
     @Override
     public CompletableFuture<Collection<Key>> list(final Key prefix) {
-        return CompletableFuture.supplyAsync(
+        return Single.fromCallable(
```
Comment (olegmoz) [inline src/main/java/com/artipie/asto/fs/FileStorage.java:94]: @Sammers21 what is the reason for this change?

File: src/main/java/com/artipie/asto/fs/FileStorage.java [hunk 3]
```diff
@@ -151,65 +127,52 @@ public FileStorage(final Path path, final ExecutorService exec) {
...
+            .flatMapCompletable(
+                file -> new RxFile(file, this.fls).save(Flowable.fromPublisher(content))
+            ).onErrorResumeNext(
+                throwable -> new RxFile(this.path(key), this.fls)
+                    .delete()
+                    .andThen(Completable.error(throwable))
+            )
+            .to(CompletableInterop.await())
+            .<Void>thenApply(o -> null)
+            .toCompletableFuture();
```
Comment (olegmoz) [inline src/main/java/com/artipie/asto/fs/FileStorage.java:150]: @Sammers21 `save` operation becomes non-atomic after these changes. If there was some value saved for the `key` before the operation it will be lost on error. Also the file could be written partially if thread/process is terminated

File: src/main/java/com/artipie/asto/fs/FileStorage.java [hunk 4]
```diff
@@ -245,32 +208,6 @@ public FileStorage(final Path path, final ExecutorService exec) {
...
-            dst -> {
-                try {
-                    Files.move(source, dst, StandardCopyOption.REPLACE_EXISTING);
-                } catch (final IOException iex) {
-                    throw new UncheckedIOException(iex);
-                }
-            },
-            this.exec
-        );
-    }
```
Comment (olegmoz) [inline src/main/java/com/artipie/asto/fs/FileStorage.java:272]: @Sammers21 this method probably should be brought back and reused in `save` and `move` methods in order to make `save` atomic again

File: src/main/java/com/artipie/asto/fs/RxFile.java [hunk 5]
```diff
@@ -160,20 +165,6 @@ public Completable delete() {
...
-                            res.onSuccess(Files.size(this.file));
-                        } catch (final IOException iex) {
-                            res.onError(iex);
-                        }
-                    }
-                );
-                return res;
-            }
-        );
+        return Single.fromCallable(() -> Files.size(this.file));
```
Comment (olegmoz) [inline src/main/java/com/artipie/asto/fs/RxFile.java:168]: @Sammers21 that is a blocking operation executing in common thread pool. Before the changes the blocking operation was performed on special thread pool

File: src/main/java/com/artipie/asto/test/TestResource.java [hunk 6]
```diff
@@ -47,12 +48,19 @@
...
+     */
+    private final FileSystem fls;
+
     /**
      * Ctor.
      * @param name Resource path
+     * @param fls The file system.
      */
-    public TestResource(final String name) {
+    public TestResource(final String name, final FileSystem fls) {
```
Comment (olegmoz) [inline src/main/java/com/artipie/asto/test/TestResource.java:61]: @Sammers21 that's really not nice to add another argument to this class, as it used in many tests. Adding another argument here will make it way harder to use

File: src/test/java/com/artipie/asto/RxFileTest.java [hunk 7]
```diff
@@ -46,10 +47,11 @@
     @Test
     @Timeout(1)
     public void rxFileFlowWorks(@TempDir final Path tmp) throws IOException {
+        final Vertx vertx = Vertx.vertx();
```
Comment (olegmoz) [inline src/test/java/com/artipie/asto/RxFileTest.java:50]: @Sammers21 `Vertx` instance is created and closed in every test of this test class. I think it will be better to manage it in `@BeforeEach`, `@AfterEach` methods

**Review by olegmoz — CHANGES_REQUESTED** [review]
@Sammers21 thanks for the PR, please see multiple comments

One way to resolve it and make the changes safer is to keep `FileStorage` as it is and create another implementation, say `VertxFileStorage`. This way we will not break things and easily switch back to previous implementation if faster implementation will have some functionality flaws

File: src/main/java/com/artipie/asto/fs/FileStorage.java (hunk 1 above)
Comment (Sammers21) [reply src/main/java/com/artipie/asto/fs/FileStorage.java:84]: @olegmoz, the `CompletableFuture.supplyAsync` call does not allow us to use vert.x as something to delegate work to

File: src/main/java/com/artipie/asto/fs/FileStorage.java (hunk 3 above)
Comment (Sammers21) [reply src/main/java/com/artipie/asto/fs/FileStorage.java:150]: @olegmoz, there is no test covering it.

Comment (Sammers21) [reply src/main/java/com/artipie/asto/fs/FileStorage.java:150]: @olegmoz, and why we would need it here?

File: src/main/java/com/artipie/asto/fs/RxFile.java (hunk 5 above)
Comment (Sammers21) [reply src/main/java/com/artipie/asto/fs/RxFile.java:168]: RxJava can also be reconfigured to use the Vert.x scheduler:

```
RxJavaPlugins.setComputationSchedulerHandler(s -> RxHelper.scheduler(vertx));
RxJavaPlugins.setIoSchedulerHandler(s -> RxHelper.blockingScheduler(vertx));
RxJavaPlugins.setNewThreadSchedulerHandler(s -> RxHelper.scheduler(vertx));
```

File: src/main/java/com/artipie/asto/test/TestResource.java (hunk 6 above)
Comment (Sammers21) [reply src/main/java/com/artipie/asto/test/TestResource.java:61]: @olegmoz, what do you suggest?

Comment (olegmoz) [reply src/main/java/com/artipie/asto/test/TestResource.java:61]: @Sammers21 I suggest to create separate class `VertxFileStorage` and keep `FileStorage` intact. We can switch to using it in Artipie without breaking other stuff

File: src/main/java/com/artipie/asto/fs/FileStorage.java (hunk 3 above)
Comment (olegmoz) [reply src/main/java/com/artipie/asto/fs/FileStorage.java:150]: @Sammers21 please read comment above, we need `save` operation to be atomic. This property is used in many adapters. That's bad we lack tests for that, it should be addressed as well

File: src/main/java/com/artipie/asto/fs/FileStorage.java (hunk 1 above)
Comment (olegmoz) [reply src/main/java/com/artipie/asto/fs/FileStorage.java:84]: @Sammers21 I am sorry but why do we need to give this kind of tasks to Vert.x? And how `Single.fromCallable` is related to Vert.x?

**Comment by Sammers21** [comment]
@olegmoz, if i do a separated implementation, what should i do with `RxFile`?

File: src/main/java/com/artipie/asto/fs/FileStorage.java (hunk 3 above)
Comment (Sammers21) [reply src/main/java/com/artipie/asto/fs/FileStorage.java:150]: @olegmoz, if this property is "used in many adapters", why there are no tests for it?

File: src/main/java/com/artipie/asto/fs/FileStorage.java (hunk 1 above)
Comment (Sammers21) [reply src/main/java/com/artipie/asto/fs/FileStorage.java:84]: @olegmoz, you can delegate tasks to vert.x blocking executor

**Comment by Sammers21** [comment]
@olegmoz, review the PR one more time please.

File: src/main/java/com/artipie/asto/fs/VertxFileStorage.java [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+            () -> {
+                try {
+                    return Files.size(this.path(key));
+                } catch (final NoSuchFileException nofile) {
+                    throw new ValueNotFoundException(key, nofile);
+                } catch (final IOException iex) {
+                    throw new UncheckedIOException(iex);
+                }
+            }
+        ).subscribeOn(RxHelper.blockingScheduler(this.vertx.getDelegate()))
```
Comment (olegmoz) [inline src/main/java/com/artipie/asto/fs/VertxFileStorage.java:209]: @Sammers21 many others methods in this class run blocking operations using `Files`. May we use same blocking scheduler for them as well?

**Review by olegmoz — CHANGES_REQUESTED** [review]
@Sammers21 thanks, please see one comment

**Comment by Sammers21** [comment]
@olegmoz, review the PR one more time, please.

**Review by olegmoz — APPROVED** [review]
@Sammers21 thanks, merging

**Comment by Sammers21** [comment]
@rultor, release tag=`0.29`

**Comment by 0crat** [comment]
Job `gh:artipie/asto#279` is not assigned, can't get performer
